boot/zephyr/main.c

```diff
@@ -175,6 +200,58 @@ static void do_boot(struct boot_rsp *rsp)
 }
 #endif
 
+#if defined(CONFIG_LOG) && !defined(CONFIG_LOG_IMMEDIATE) &&\
+    !defined(CONFIG_LOG_PROCESS_THREAD)
+/* The log internal thread for log processing can't transfer log well as has too
+ * low priority.
+ * Dedicated thread for log processing below uses highest application
+ * priority. This allows to transmit all logs without adding k_sleep/k_yield
+ * anywhere else int the code.
+ */
+
+/* most simple log processing theread */
+void boot_log_thread_func(void *dummy1, void *dummy2, void *dummy3)
+{
+    (void)dummy1;
+    (void)dummy2;
+    (void)dummy3;
+
+     log_init();
+
+     while (1) {
+             if (log_process(false) == false) {
+                    if (boot_log_stop) {
+                        boot_log_stop = false;
+                        break;
+                    }
+                    k_sleep(BOOT_LOG_PROCESSING_INTERVAL);
+             }
+     }
+}
+
+void zephyr_boot_log_start(void)
+{
+        /* start logging thread */
+        k_thread_create(&boot_log_thread, boot_log_stack,
+                K_THREAD_STACK_SIZEOF(boot_log_stack),
+                boot_log_thread_func, NULL, NULL, NULL,
+                K_HIGHEST_APPLICATION_THREAD_PRIO, 0,
+                BOOT_LOG_PROCESSING_INTERVAL);
+
+        k_thread_name_set(&boot_log_thread, "logging");
+}
+
+void zephyr_boot_log_stop(void)
+{
+    boot_log_stop = true;
+
+    /* busy wait until log procesing thread expired*/
+    while (boot_log_stop) {
```

Comment:
Are you sure this is enough? Might be worth using k_kthread_abort() afterwards:

https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#_CPPv414k_thread_abort7k_tid_t

(I just am being paranoid about cleaning up according to Zephyr API promises, rather than what happens to work today.)

Reply:
quote [thread termination behavior description](
https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#thread-termination):
> Once a thread is started it typically executes forever. However, a thread may synchronously end its execution by returning from its entry point function. This is known as termination.

In the next paragraph (about aborting):
> it is typically preferable to signal a thread to terminate itself gracefully, rather than aborting it.

So thread ends execution and that all needed. I might call abort but according to this, it is not  required.

Reply:
Right, but the current implementation looks racy. How do you know the thread is done, instead of just having set the variable and gotten scheduled out? Are all the threads cooperative?

Reply:
Looking at https://github.com/zephyrproject-rtos/zephyr/issues/21500, it seems like the common pattern right now is to use a semaphore given at the end of the exiting thread, since there is no `k_thread_join()` in Zephyr.

 From that issue:

> There's no good way to block until another thread exits.
> 
> Our test cases currently manage this in a few ways:
>
>    1. Create an equal priority co-op thread, or higher priority thread, and then yield() the CPU, or sleep for a while. Assuming the child thread never blocks and then exits, control returns after yield() completes...if we are in uniprocessor. In SMP the child could spawn on another CPU in which case after the yield() returns the child is probably still running. The workaround is the 1cpu variant of test case declarations, which spawn a busy-looping co-op thread on all other CPUs.
>    2. Synchronize by having the parent take a semaphore given by the child at the end of its entry function. Potentially race-prone if the child hasn't completely exited and the parent tries to re-use the thread object, resulting in a crash, but still safer than #1.

So I think on Nordic SoCs, the current implementation, which uses a higher priority thread and a flag global, is probably safe. But just for the sake of keeping to the common pattern, how about using approach 2. for now (give a semaphore at log thread exit), and add a comment to switch to using `k_thread_join()` as soon as that is available?

Reply:
^^ updated.
I could even use the same semaphore (but numeric one) to request thread termination - but I fell this as a overkill, so kept trigger via volatile variable,